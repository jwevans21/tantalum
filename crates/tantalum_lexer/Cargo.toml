[package]
name = "tantalum_lexer"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[dev-dependencies]
insta = { workspace = true}
pretty_assertions = { workspace = true }
rstest = { workspace = true }


[lints]
workspace = true
