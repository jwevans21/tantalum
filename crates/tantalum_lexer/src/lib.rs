//! # Tantalum Lexer
//!
//! This crate provides a lexer for the Tantalum language.

use tantalum_source::{SourceFile, SourceLocation, SourceSpan};
use tantalum_syntax::{SyntaxKind, SyntaxToken};

/// A lexer for the Tantalum language.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Lexer<'file> {
    /// The source file that the lexer is currently lexing.
    file: &'file SourceFile,
    /// The current location in the source file.
    location: SourceLocation,
}

impl<'file> Lexer<'file> {
    /// Create a new lexer from a source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let lexer = Lexer::new(file);
    /// ```
    #[must_use]
    pub fn new(file: &'file SourceFile) -> Self {
        Self {
            file,
            location: file.location_start(),
        }
    }

    /// Get the source file that the lexer is currently lexing.
    #[must_use]
    pub fn source(&self) -> &'file SourceFile {
        self.file
    }

    /// Get the current location in the source file.
    #[must_use]
    pub fn location(&self) -> SourceLocation {
        self.location
    }

    /// Get the character at a location in the source file.
    fn at_location(&self, location: SourceLocation) -> Option<char> {
        self.file.read_char(location)
    }

    /// Extract the next token from the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.next_token().expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    /// ```
    #[must_use]
    pub fn next_token(&mut self) -> Option<SyntaxToken> {
        let start = self.location;

        let (kind, end) = self.characters_to_token(start)?;

        self.location = end;

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Peek at the next token in the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.peek_token(1).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    ///
    /// let token = lexer.peek_token(2).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Whitespace);
    ///
    /// let token = lexer.peek_token(3).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Identifier);
    /// ```
    #[must_use]
    pub fn peek_token(&self, mut n: usize) -> Option<SyntaxToken> {
        let mut start = self.location;

        let (kind, end) = loop {
            n -= 1;
            let (kind, end) = self.characters_to_token(start)?;

            if n == 0 {
                break (kind, end);
            }

            start = end;
        };

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Convert a character to a [`SyntaxKind`].
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    fn characters_to_token(&self, start: SourceLocation) -> Option<(SyntaxKind, SourceLocation)> {
        let mut current = start;

        let c = self.at_location(current)?;
        current = current.next_by(c.len_utf8());

        let token = match c {
            '(' => SyntaxKind::LParen,
            ')' => SyntaxKind::RParen,
            '{' => SyntaxKind::LBrace,
            '}' => SyntaxKind::RBrace,
            '[' => SyntaxKind::LBracket,
            ']' => SyntaxKind::RBracket,
            '<' => SyntaxKind::LAngle,
            '>' => SyntaxKind::RAngle,
            ',' => SyntaxKind::Comma,
            ':' => SyntaxKind::Colon,
            ';' => SyntaxKind::Semicolon,
            '.' => SyntaxKind::Dot,
            '&' => SyntaxKind::Ampersand,
            '|' => SyntaxKind::Pipe,
            '^' => SyntaxKind::Caret,
            '+' => SyntaxKind::Plus,
            '-' => SyntaxKind::Minus,
            '*' => SyntaxKind::Star,
            '/' => SyntaxKind::Slash,
            '\\' => SyntaxKind::Backslash,
            '%' => SyntaxKind::Percent,
            '=' => SyntaxKind::Equals,
            '!' => SyntaxKind::Bang,
            '~' => SyntaxKind::Tilde,
            '?' => SyntaxKind::Question,
            '@' => SyntaxKind::At,
            '0'..='9' => SyntaxKind::Digit,
            '_' | 'a'..='z' | 'A'..='Z' => {
                while let Some(c) = self.at_location(current) {
                    match c {
                        '_' | 'a'..='z' | 'A'..='Z' | '0'..='9' => {
                            current = current.next_by(c.len_utf8());
                        }
                        _ => break,
                    }
                }

                match self.file.read_span(SourceSpan::new(
                    self.file.id(),
                    self.location.offset(),
                    current.offset(),
                ))? {
                    "fn" => SyntaxKind::Fn,
                    "let" => SyntaxKind::Let,
                    "return" => SyntaxKind::Return,
                    "if" => SyntaxKind::If,
                    "else" => SyntaxKind::Else,
                    "true" => SyntaxKind::True,
                    "false" => SyntaxKind::False,
                    _ => SyntaxKind::Identifier,
                }
            }
            c if c.is_whitespace() => SyntaxKind::Whitespace,
            _ => SyntaxKind::Other,
        };

        Some((token, current))
    }
}

#[test]
fn keywords() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "fn let return if else true false".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 7);

    assert_eq!(tokens[0].kind(), SyntaxKind::Fn);
    assert_eq!(tokens[1].kind(), SyntaxKind::Let);
    assert_eq!(tokens[2].kind(), SyntaxKind::Return);
    assert_eq!(tokens[3].kind(), SyntaxKind::If);
    assert_eq!(tokens[4].kind(), SyntaxKind::Else);
    assert_eq!(tokens[5].kind(), SyntaxKind::True);
    assert_eq!(tokens[6].kind(), SyntaxKind::False);
}

#[test]
fn identifiers() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "a A _ a1 A1 _1 _a _A __AA01".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 9);

    assert_eq!(tokens[0].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[1].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[2].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[3].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[4].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[5].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[6].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[7].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[8].kind(), SyntaxKind::Identifier);
}
