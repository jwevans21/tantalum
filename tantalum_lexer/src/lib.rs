//! # Tantalum Lexer
//!
//! Consumes source code written in the tantalum language and produces tokens
//! to be consumed by a parser.

#[cfg(test)]
mod tests;

pub mod token;
pub mod token_kind;

use tantalum_span::{Location, Span, Spanned};

use crate::token::Token;
use crate::token_kind::TokenKind;

/// The lexer for the Tantalum language, this will consume source code and produce
/// tokens with origin information. This lexer is designed to be used in a streaming
/// fashion, where tokens are consumed as they are needed.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Lexer<'file_name, 'source> {
    /// The current location of the lexer
    location: Location<'file_name>,
    /// The source code to lexed into tokens
    source: &'source str,
}

impl<'file_name, 'source> Lexer<'file_name, 'source> {
    /// Create a lexer for the Tantalum language over the provided `source`
    ///
    /// The file name is provided to allow useful diagnostics to be produced
    ///
    /// # Example
    /// ```
    /// use tantalum_lexer::Lexer;
    ///
    /// let lexer = Lexer::new("main.tan", "fn main(): i32 { return 0; }");
    ///
    /// // Use the lexer here (will likely require `lexer` to be mutable)
    /// ```
    #[must_use]
    #[inline]
    pub fn new(file_name: &'file_name str, source: &'source str) -> Self {
        return Self {
            location: Location::new(file_name),
            source,
        };
    }

    #[must_use]
    #[inline]
    pub fn location(&self) -> Location<'file_name> {
        return self.location;
    }

    #[must_use]
    #[inline]
    pub fn source(&self) -> &'source str {
        return self.source;
    }

    #[must_use]
    #[inline]
    pub fn file_name(&self) -> &'file_name str {
        return self.location.file_name();
    }

    #[must_use]
    #[inline]
    #[expect(clippy::too_many_lines, reason = "Long match statement")]
    pub fn next_token(&mut self) -> Option<Spanned<'file_name, Token<'source>>> {
        self.skip_whitespace();

        let start = self.location;

        /// Produce a Token based on a type and length
        macro_rules! lex {
            // Special case  for tokens already consumed
            ($production:path, 0) => {{
                return self.create_token($production, start);
            }};
            // General case for tokens that need to be consumed
            ($production:path, $length:expr) => {{
                let _: char = self.next_characters($length)?;
                return self.create_token($production, start);
            }};
        }

        match self.peek_characters(1)? {
            '(' => lex!(TokenKind::LeftParen, 1),
            ')' => lex!(TokenKind::RightParen, 1),
            '{' => lex!(TokenKind::LeftBrace, 1),
            '}' => lex!(TokenKind::RightBrace, 1),
            '[' => lex!(TokenKind::LeftBracket, 1),
            ']' => lex!(TokenKind::RightBracket, 1),
            ',' => lex!(TokenKind::Comma, 1),
            ';' => lex!(TokenKind::Semicolon, 1),

            ':' => {
                if let Some(':') = self.peek_characters(2) {
                    lex!(TokenKind::ColonColon, 2)
                }
                lex!(TokenKind::Colon, 1)
            }
            '.' => match self.peek_characters(2) {
                Some('*') => lex!(TokenKind::DotStar, 2),
                Some('&') => lex!(TokenKind::DotAmpersand, 2),
                Some('.')
                    if self
                        .peek_characters(3)
                        .is_some_and(|character| return character == '.') =>
                {
                    lex!(TokenKind::DotDotDot, 3)
                }
                _ => lex!(TokenKind::Dot, 1),
            },
            '=' => {
                if let Some('=') = self.peek_characters(2) {
                    lex!(TokenKind::EqualEqual, 2)
                }
                lex!(TokenKind::Equal, 1)
            }
            '+' => lex!(TokenKind::Plus, 1),
            '-' => lex!(TokenKind::Minus, 1),
            '*' => lex!(TokenKind::Star, 1),
            '/' => lex!(TokenKind::Slash, 1),
            '%' => lex!(TokenKind::Percent, 1),
            '&' => {
                if let Some('&') = self.peek_characters(2) {
                    lex!(TokenKind::AmpersandAmpersand, 2)
                }
                lex!(TokenKind::Ampersand, 1)
            }
            '|' => {
                if let Some('|') = self.peek_characters(2) {
                    lex!(TokenKind::PipePipe, 2)
                }
                lex!(TokenKind::Pipe, 1)
            }
            '!' => {
                if let Some('=') = self.peek_characters(2) {
                    lex!(TokenKind::ExclamationEqual, 2)
                }
                lex!(TokenKind::Exclamation, 1)
            }
            '^' => lex!(TokenKind::Caret, 1),
            '~' => lex!(TokenKind::Tilde, 1),
            '<' => match self.peek_characters(2) {
                Some('<') => lex!(TokenKind::LeftAngleLeftAngle, 2),
                Some('=') => lex!(TokenKind::LeftAngleEqual, 2),
                _ => lex!(TokenKind::LeftAngle, 1),
            },
            '>' => match self.peek_characters(2) {
                Some('>') => lex!(TokenKind::RightAngleRightAngle, 2),
                Some('=') => lex!(TokenKind::RightAngleEqual, 2),
                _ => lex!(TokenKind::RightAngle, 1),
            },

            'a'..='z' | 'A'..='Z' | '_' => {
                while let Some(character) = self.peek_characters(1) {
                    if character.is_ascii_alphanumeric() || character == '_' {
                        let _: char = self.next_characters(1)?;
                    } else {
                        break;
                    }
                }

                match self.source.get(start.range(&self.location))? {
                    "fn" => lex!(TokenKind::KeywordFn, 0),
                    "extern" => lex!(TokenKind::KeywordExtern, 0),
                    "let" => lex!(TokenKind::KeywordLet, 0),
                    "if" => lex!(TokenKind::KeywordIf, 0),
                    "else" => lex!(TokenKind::KeywordElse, 0),
                    "while" => lex!(TokenKind::KeywordWhile, 0),
                    "for" => lex!(TokenKind::KeywordFor, 0),
                    "return" => lex!(TokenKind::KeywordReturn, 0),
                    "break" => lex!(TokenKind::KeywordBreak, 0),
                    "continue" => lex!(TokenKind::KeywordContinue, 0),
                    "const" => lex!(TokenKind::KeywordConst, 0),
                    "true" => lex!(TokenKind::KeywordTrue, 0),
                    "false" => lex!(TokenKind::KeywordFalse, 0),
                    _ => lex!(TokenKind::Identifier, 0),
                }
            }

            '0'..='9' => {
                // Check for binary, octal, decimal, or hexadecimal integer literals
                match self.peek_characters(2) {
                    Some('b')
                        if self
                            .peek_characters(3)
                            .is_some_and(|character| return character.is_digit(2)) =>
                    {
                        let _: char = self.next_characters(2)?;

                        while let Some(character) = self.peek_characters(1) {
                            if character.is_digit(2) {
                                let _: char = self.next_characters(1)?;
                            } else {
                                break;
                            }
                        }

                        lex!(TokenKind::BinaryIntegerLiteral, 0);
                    }
                    Some('o')
                        if self
                            .peek_characters(3)
                            .is_some_and(|character| return character.is_digit(8)) =>
                    {
                        let _: char = self.next_characters(2)?;

                        while let Some(character) = self.peek_characters(1) {
                            if character.is_digit(8) {
                                let _: char = self.next_characters(1)?;
                            } else {
                                break;
                            }
                        }

                        lex!(TokenKind::OctalIntegerLiteral, 0);
                    }
                    Some('x')
                        if self
                            .peek_characters(3)
                            .is_some_and(|character| return character.is_ascii_hexdigit()) =>
                    {
                        let _: char = self.next_characters(2)?;

                        while let Some(character) = self.peek_characters(1) {
                            if character.is_ascii_hexdigit() {
                                let _: char = self.next_characters(1)?;
                            } else {
                                break;
                            }
                        }

                        lex!(TokenKind::HexadecimalIntegerLiteral, 0);
                    }
                    _ => {
                        while let Some(character) = self.peek_characters(1) {
                            if character.is_ascii_digit() {
                                let _: char = self.next_characters(1)?;
                            } else {
                                break;
                            }
                        }

                        match self.peek_characters(1) {
                            Some('.')
                                if self
                                    .peek_characters(2)
                                    .is_some_and(|character| character.is_ascii_digit()) =>
                            {
                                let _: char = self.next_characters(1)?;

                                while let Some(character) = self.peek_characters(1) {
                                    if character.is_ascii_digit() {
                                        let _: char = self.next_characters(1)?;
                                    } else {
                                        break;
                                    }
                                }

                                match self.peek_characters(1) {
                                    Some('e' | 'E') => {
                                        let _: char = self.next_characters(1)?;

                                        if let Some('+' | '-') = self.peek_characters(1) {
                                            let _: char = self.next_characters(1)?;
                                        }

                                        while let Some(character) = self.peek_characters(1) {
                                            if character.is_ascii_digit() {
                                                let _: char = self.next_characters(1)?;
                                            } else {
                                                break;
                                            }
                                        }

                                        lex!(TokenKind::FloatLiteral, 0);
                                    }
                                    _ => lex!(TokenKind::FloatLiteral, 0),
                                }
                            }
                            _ => lex!(TokenKind::DecimalIntegerLiteral, 0),
                        }
                    }
                }
            }

            '"' => {
                let _: char = self.next_characters(1)?;

                while let Some(character) = self.peek_characters(1) {
                    match character {
                        '\\' => {
                            let _: char = self.next_characters(1)?;
                            let _: char = self.next_characters(1)?;
                        }
                        '"' => {
                            let _: char = self.next_characters(1)?;
                            break;
                        }
                        _ => {
                            let _: char = self.next_characters(1)?;
                        }
                    }
                }

                lex!(TokenKind::StringLiteral, 0);
            }

            '\'' => {
                let _: char = self.next_characters(1)?;

                match self.peek_characters(1) {
                    Some('\\') => {
                        let _: char = self.next_characters(1)?;
                        let _: char = self.next_characters(1)?;
                    }
                    _ => {
                        let _: char = self.next_characters(1)?;
                    }
                }

                match self.peek_characters(1) {
                    Some('\'') => {
                        let _: char = self.next_characters(1)?;
                    }
                    _ => {
                        return None;
                    }
                }

                lex!(TokenKind::CharacterLiteral, 0);
            }

            _ => lex!(TokenKind::Unknown, 1),
        }
    }

    /// Build a token with the current state based on a token type and the
    /// starting position of the token
    fn create_token(
        &self,
        token_kind: TokenKind,
        start: Location<'file_name>,
    ) -> Option<Spanned<'file_name, Token<'source>>> {
        let span = Span::new(start, self.location);
        return Some(Spanned::new(
            span,
            Token::new(self.source.get(span.range())?, token_kind),
        ));
    }

    /// Skip any whitespace characters in the source code
    fn skip_whitespace(&mut self) {
        while let Some(character) = self.peek_characters(1) {
            if character.is_whitespace() {
                let _: char = self
                    .next_characters(1)
                    .expect("Failed to unwrap character already shown to exist");
            } else {
                break;
            }
        }
    }

    /// Peek into the source code by `count` characters
    ///
    /// Return the the `count` character from the current position if it exists,
    /// otherwise it returns `None`
    #[must_use]
    #[inline]
    fn peek_characters(&self, count: usize) -> Option<char> {
        let next = self.source.get(self.location.position()..)?;

        if next.len() < count {
            return None;
        }

        return next.chars().take(count).last();
    }

    /// Advance the lexer by `count` characters
    ///
    /// If `count` exceeds the remaining length then `None` is returned and
    /// the `Lexer` has not been updated.
    ///
    /// This will update the position based on UTF-8 character lengths, it will
    /// also update the line and column as appropriate
    #[must_use]
    #[inline]
    fn next_characters(&mut self, count: usize) -> Option<char> {
        let next = self.source.get(self.location.position()..)?;

        if next.len() < count {
            return None;
        }

        return next
            .chars()
            .take(count)
            .inspect(|character| {
                self.location.advance(*character);
            })
            .last();
    }
}

impl<'file_name, 'source> Iterator for Lexer<'file_name, 'source> {
    type Item = Spanned<'file_name, Token<'source>>;

    #[inline]
    fn next(&mut self) -> Option<Self::Item> {
        return self.next_token();
    }
}
