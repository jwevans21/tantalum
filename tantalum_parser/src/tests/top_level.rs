use tantalum_lexer::Lexer;

use crate::Parser;

#[test]
fn function_declaration() {
    let source = r"fn main() {}";
    let mut parser = Parser::new(Lexer::new("function_declaration", source));

    let result = parser.parse_top_level();

    insta::assert_ron_snapshot!(result);
}

#[test]
fn function_declaration_with_parameters() {
    let source = r"fn main(argc: i32, argv: [*const u8]) {}";
    let mut parser = Parser::new(Lexer::new("function_declaration_with_parameters", source));

    let result = parser.parse_top_level();

    insta::assert_ron_snapshot!(result);
}

#[test]
fn function_declaration_with_return_type() {
    let source = r"fn main(): i32 {}";
    let mut parser = Parser::new(Lexer::new("function_declaration_with_return_type", source));

    let result = parser.parse_top_level();

    insta::assert_ron_snapshot!(result);
}

#[test]
fn external_function_declaration() {
    let source = r"extern fn puts(s: *const u8): i32;";
    let mut parser = Parser::new(Lexer::new("external_function_declaration", source));

    let result = parser.parse_top_level();

    insta::assert_ron_snapshot!(result);
}

#[test]
fn external_function_declaration_with_variadic() {
    let source = r"extern fn printf(format: *const u8, ...): i32;";
    let mut parser = Parser::new(Lexer::new(
        "external_function_declaration_with_variadic",
        source,
    ));

    let result = parser.parse_top_level();

    insta::assert_ron_snapshot!(result);
}
