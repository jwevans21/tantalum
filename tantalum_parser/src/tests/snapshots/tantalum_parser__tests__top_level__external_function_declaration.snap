---
source: tantalum_parser/src/tests/top_level.rs
expression: result
snapshot_kind: text
---
Ok(Spanned(
  span: Span(
    start: Location(
      file_name: "external_function_declaration",
      position: 0,
      line: 1,
      column: 1,
    ),
    end: Location(
      file_name: "external_function_declaration",
      position: 34,
      line: 1,
      column: 35,
    ),
  ),
  data: ExternalFunction(ExternalFunction(
    name: Spanned(
      span: Span(
        start: Location(
          file_name: "external_function_declaration",
          position: 10,
          line: 1,
          column: 11,
        ),
        end: Location(
          file_name: "external_function_declaration",
          position: 14,
          line: 1,
          column: 15,
        ),
      ),
      data: "puts",
    ),
    parameters: Spanned(
      span: Span(
        start: Location(
          file_name: "external_function_declaration",
          position: 14,
          line: 1,
          column: 15,
        ),
        end: Location(
          file_name: "external_function_declaration",
          position: 28,
          line: 1,
          column: 29,
        ),
      ),
      data: [
        Spanned(
          span: Span(
            start: Location(
              file_name: "external_function_declaration",
              position: 15,
              line: 1,
              column: 16,
            ),
            end: Location(
              file_name: "external_function_declaration",
              position: 27,
              line: 1,
              column: 28,
            ),
          ),
          data: Named(NamedParameter(
            name: Spanned(
              span: Span(
                start: Location(
                  file_name: "external_function_declaration",
                  position: 15,
                  line: 1,
                  column: 16,
                ),
                end: Location(
                  file_name: "external_function_declaration",
                  position: 16,
                  line: 1,
                  column: 17,
                ),
              ),
              data: "s",
            ),
            ty: Spanned(
              span: Span(
                start: Location(
                  file_name: "external_function_declaration",
                  position: 18,
                  line: 1,
                  column: 19,
                ),
                end: Location(
                  file_name: "external_function_declaration",
                  position: 27,
                  line: 1,
                  column: 28,
                ),
              ),
              data: Pointer(PointerType(
                ty: Spanned(
                  span: Span(
                    start: Location(
                      file_name: "external_function_declaration",
                      position: 19,
                      line: 1,
                      column: 20,
                    ),
                    end: Location(
                      file_name: "external_function_declaration",
                      position: 27,
                      line: 1,
                      column: 28,
                    ),
                  ),
                  data: Const(ConstType(
                    ty: Spanned(
                      span: Span(
                        start: Location(
                          file_name: "external_function_declaration",
                          position: 25,
                          line: 1,
                          column: 26,
                        ),
                        end: Location(
                          file_name: "external_function_declaration",
                          position: 27,
                          line: 1,
                          column: 28,
                        ),
                      ),
                      data: Named(NamedType(
                        name: Spanned(
                          span: Span(
                            start: Location(
                              file_name: "external_function_declaration",
                              position: 25,
                              line: 1,
                              column: 26,
                            ),
                            end: Location(
                              file_name: "external_function_declaration",
                              position: 27,
                              line: 1,
                              column: 28,
                            ),
                          ),
                          data: "u8",
                        ),
                      )),
                    ),
                  )),
                ),
              )),
            ),
          )),
        ),
      ],
    ),
    return_type: Some(Spanned(
      span: Span(
        start: Location(
          file_name: "external_function_declaration",
          position: 30,
          line: 1,
          column: 31,
        ),
        end: Location(
          file_name: "external_function_declaration",
          position: 33,
          line: 1,
          column: 34,
        ),
      ),
      data: Named(NamedType(
        name: Spanned(
          span: Span(
            start: Location(
              file_name: "external_function_declaration",
              position: 30,
              line: 1,
              column: 31,
            ),
            end: Location(
              file_name: "external_function_declaration",
              position: 33,
              line: 1,
              column: 34,
            ),
          ),
          data: "i32",
        ),
      )),
    )),
  )),
))
